### DISCUSSION TITLE
MSBuild throwing our error WIX0822 The variable 'Platform' with value 'AnyCPU'...

### DISCUSSION THREAD

**Discussion by ColinWPA** [wixtoolset/issues#8124] [Questions] [answered]
My wixproj builds perfectly inside VS2022, but using MSBuild on the project falls over and reports:

wix.exe : error WIX0288: The variable 'Platform' with value 'AnyCPU' was previously declared with value 'x86'. [C:\Workspace\90504_SmartCellConfigTool\Trunk\SetupSmartCellConfigTool\SetupSmartCell
ConfigTool.wixproj]

MY project file specifies the the platform must be x86 and configuration release viz:

<Project Sdk="WixToolset.Sdk/5.0.0" ToolsVersion="5.0">
  <PropertyGroup>
    <DefineConstants>Configuration=Release</DefineConstants>
    <DefineConstants>Platform=x86</DefineConstants>
  </PropertyGroup>
  <PropertyGroup>
    <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|x86' " />
  <PropertyGroup>
    <MaximumVersion>99.99.9999</MaximumVersion>
    <ExtVersion>1.2.102</ExtVersion>
    <ExtSourceFolder>C:\Workspace\90504_SmartCellConfigTool\Trunk\EMS Config Tool\bin\Release</ExtSourceFolder>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x86'">
    <DefineConstants>Version=$(ExtVersion);SourceFolder=$(ExtSourceFolder)</DefineConstants>
  </PropertyGroup>
.... and so on.

So how does the platform get changed to AnyCPU - that is not what I want nor what I have specified.
Prior to switching to v5.0 this did work (in 4.0.4) so is there another option needed on the msbuild command line

Any suggestions to resolve this please?

**Comment by robmen** [comment]
MSBuild your .sln and you'll likely get the same behavior as VS. A .sln adds a lot of whacky additional context, especially if you configured it via the Build Configuration Manager in VS.

Personally, I try to avoid the differences .sln add but that's more about MSBuild than WiX.

**Reply by ColinWPA** [reply]
Thanks for the suggestion. So I recreated my wixproj without any .sln file involved, and tried it again, this time specifying the .wiproj file to MSBuild. It is still throwing the same error of 

wix.exe : error WIX0288: The variable 'Platform' with value 'AnyCPU' was previously declared with value 'x86'. [C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\SetupSmartCellConfigTool.wixproj]

I can see that "somewhere" the Platform variable is getting set to "AnyCPU" (but where I don't know) and the error is thrown when my wixproj file tries to set it the required "x86" value,,,,, Note: there is no .sln file in the path. It is almost as if the wix.exe file defaults to "AnyCPU" and doesn't want it to be anything else? Maybe I'm missing some relevant point that makes changing the platform to a value different to the default possible?

**Reply by ColinWPA** [reply]
invocation: msbuild -p:ExtVersion="1.0.2" -p:ExtSourceFolder="C:\\Workspace\\90504_SmartCellConfigTool\\Trunk\\EMS Config Tool\\bin\\Release -ignore:.sln -t:x86 SetupSmartCellConfigTool.wixproj

**Comment by ColinWPA** [comment]
Still no joy here, to my mind this is a wix.exe issue that probably needs a setting to be passed to it so as to not try to build for all possibly configurations when only one is required. Here is the full trace of the problem:

C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller>msbuild -p:ExtVersion=1.2.102 SetupSmartCellConfigTool.wixproj
MSBuild version 17.9.8+b34f75857 for .NET Framework
Build started 2024/04/17 08:22:57.

Project "C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\SetupSmartCellConfigTool.wixproj" on node 1 (default targets).
CoreCompile:
  C:\Users\attwelc\.nuget\packages\wixtoolset.sdk\5.0.0\tools\net472\x86\wix.exe build -platform x86 -out obj\Debug\SetupSmartCellConfigTool.msi -outputType Package -pdb obj\Debug\SetupSmartCellConfigTool.wixpdb -pdbType full -culture n
  ull -d Platform=x86 -d Configuration=Debug -d OutDir=bin\Debug\ -d InstallerPlatform=x86 -d Platform=AnyCPU -d ProjectDir=C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\ -d ProjectExt=.wixproj -d ProjectFileName=Setup
  SmartCellConfigTool.wixproj -d ProjectName=SetupSmartCellConfigTool -d ProjectPath=C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\SetupSmartCellConfigTool.wixproj -d TargetDir=C:\Workspace\90504_SmartCellConfigTool\Tr
  unk\SmartCellInstaller\bin\Debug\ -d TargetExt=.msi -d TargetFileName=SetupSmartCellConfigTool.msi -d TargetName=SetupSmartCellConfigTool -d TargetPath=C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\bin\Debug\SetupSma
  rtCellConfigTool.msi -ext C:\Users\attwelc\.nuget\packages\wixtoolset.util.wixext\5.0.0\wixext5\WixToolset.Util.wixext.dll -ext C:\Users\attwelc\.nuget\packages\wixtoolset.ui.wixext\5.0.0\wixext5\WixToolset.UI.wixext.dll -ext C:\Users
  \attwelc\.nuget\packages\wixtoolset.netfx.wixext\5.0.0\wixext5\WixToolset.Netfx.wixext.dll -intermediatefolder obj\Debug\ -trackingfile obj\Debug\SetupSmartCellConfigTool.wixproj.BindTracking-neutral.txt -nologo  Folders.wxs GroupRefs
  .wxs Groups.wxs Product.wxs RemoveFiles.wxs
wix.exe : error WIX0288: The variable 'Platform' with value 'AnyCPU' was previously declared with value 'x86'. [C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\SetupSmartCellConfigTool.wixproj]
Done Building Project "C:\Workspace\90504_SmartCellConfigTool\Trunk\SmartCellInstaller\SetupSmartCellConfigTool.wixproj" (default targets) -- FAILED.

Clearly, wix.exe has been given a bunch of parameters trying to build everything it could possibly build, but not required. 
Can anybody shed any light what the possible parameter settings to pass into the MSBuild command to resolve this. 

-NOTE- This is no .sln file anywhere in the folder that could possibly be affecting this.

Thanks in advance.

**Comment by robmen** [comment]
You are doubly defining the Platform preprocessor variable on the command line.

```
build -platform x86 -out obj\Debug\SetupSmartCellConfigTool.msi -outputType Package -pdb obj\Debug\SetupSmartCellConfigTool.wixpdb -pdbType full -culture n
ull -d Platform=x86 -d Configuration=Debug -d OutDir=bin\Debug\ -d InstallerPlatform=x86 -d Platform=AnyCPU
```

Look in you project file to find it.

**Reply by ColinWPA** [reply]
I wish it were so.... I just type MSBuild on the command line in the project folder - nothing else!

So I've tried to simplify things as much as possible with another simple project.

This is the wixproj file. There is NO .sln file

`<?xml version="1.0"?>
<Project Sdk="WixToolset.Sdk/5.0.0" ToolsVersion="5.0">
    <PropertyGroup>
        <DefineConstants>Configuration=Release</DefineConstants>
        <DefineConstants>Platform=x86</DefineConstants>
    </PropertyGroup>
    <PropertyGroup>
        <EnableDefaultCompileItems>False</EnableDefaultCompileItems>
    </PropertyGroup>
    <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'" />
    <PropertyGroup>
        <MaximumVersion>99.99.9999</MaximumVersion>
        <ExtVersion>1.0.001</ExtVersion>
        <ExtSourceFolder>C:\Workspace\EMSInstallBuilder\EMSInstallBuilder\bin\x86\Release\net8.0-windows7.0</ExtSourceFolder>
    </PropertyGroup>
    <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x86'">
        <DefineConstants>Version=$(ExtVersion);SourceFolder=$(ExtSourceFolder)</DefineConstants>
    </PropertyGroup>
    <ItemGroup>
        <Compile Include="Folders.wxs" />
        <Compile Include="GroupRefs.wxs" />
        <Compile Include="Groups.wxs" />
        <Compile Include="Product.wxs" />
        <Compile Include="RemoveFiles.wxs" />
    </ItemGroup>
    <ItemGroup>
        <PackageReference Include="WixToolset.Util.wixext" Version="5.0.0" />
        <PackageReference Include="WixToolset.UI.wixext" Version="5.0.0" />
        <PackageReference Include="WixToolset.NetFx.wixext" Version="5.0.0" />
    </ItemGroup>
</Project>`

and this is the result:

C:\Workspace\EMSInstallBuilder\InstallBuilderSetup>msbuild
MSBuild version 17.9.8+b34f75857 for .NET Framework
Build started 2024/04/17 14:06:31.

Project "C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj" on node 1 (default targets).
PrepareForBuild:
  Creating directory "bin\Debug\".
  Creating directory "obj\Debug\".
CoreCompile:
  C:\Users\attwelc\.nuget\packages\wixtoolset.sdk\5.0.0\tools\net472\x86\wix.exe build -platform x86 -out obj\Debug\Project.msi -outputType Package -pdb obj\Debug\Project.wixpdb -pdbType full -culture null -d Platform=x86 -d Configurati
  on=Debug -d OutDir=bin\Debug\ -d InstallerPlatform=x86 -d Platform=AnyCPU -d ProjectDir=C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\ -d ProjectExt=.wixproj -d ProjectFileName=Project.wixproj -d ProjectName=Project -d ProjectPat
  h=C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj -d TargetDir=C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\bin\Debug\ -d TargetExt=.msi -d TargetFileName=Project.msi -d TargetName=Project -d TargetPath=C:\Wor
  kspace\EMSInstallBuilder\InstallBuilderSetup\bin\Debug\Project.msi -ext C:\Users\attwelc\.nuget\packages\wixtoolset.util.wixext\5.0.0\wixext5\WixToolset.Util.wixext.dll -ext C:\Users\attwelc\.nuget\packages\wixtoolset.ui.wixext\5.0.0\
  wixext5\WixToolset.UI.wixext.dll -ext C:\Users\attwelc\.nuget\packages\wixtoolset.netfx.wixext\5.0.0\wixext5\WixToolset.Netfx.wixext.dll -intermediatefolder obj\Debug\ -trackingfile obj\Debug\Project.wixproj.BindTracking-neutral.txt -
  nologo  Folders.wxs GroupRefs.wxs Groups.wxs Product.wxs RemoveFiles.wxs
wix.exe : error WIX0288: The variable 'Platform' with value 'AnyCPU' was previously declared with value 'x86'. [C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj]
Done Building Project "C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj" (default targets) -- FAILED.


Build FAILED.

"C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj" (default target) (1) ->
(CoreCompile target) ->
  wix.exe : error WIX0288: The variable 'Platform' with value 'AnyCPU' was previously declared with value 'x86'. [C:\Workspace\EMSInstallBuilder\InstallBuilderSetup\Project.wixproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.56

Is there any way to get wix.exe to NOT try to build for AnyCPU? Is there any command one can write after MSBuild to ensure wix.exe does what I believe I ask it to do?

I'm trying to get this toolset approved for use by the corporation that employs me, but without being able to demonstrate that it works, I'm having a hard time doing so.....I'm fairly sure I could convince them to go for the paid-for-support option, once I can prove it is fit for purpose, Maybe it is I that am not clever enough to figure things out, but it would be a sad day to have to revert to the likes of InstallAware (from whence we came) or InstallShield, but to get on with my "real" job, I need to get done here, or move on.

Thanks for your support though, I do appreciate it.

**Reply by robmen** [reply]
You have hard-coded the preprocessor variable `Platform` in your .wixproj.

```xml
<DefineConstants>Platform=x86</DefineConstants> 
```

Remove that line, and this error will go away.

Note: If you [paid for support](https://www.firegiant.com/services/support-programs/), you'd first get an onboarding review to help align your code for success. Then, you'd get detailed answers to all your questions that would explain the ins and outs of each option.

Right now, I'm just driving by and can see the problems quickly. Free support is, well, free.

**Reply by ColinWPA** [reply]
Thank you. So how then does the correct configuration and platform get set? You'll see further down the project the test to see if the configuration and platform is set to the correct values, so that the version and source file parameters can be set.... Note that this code was lifted straight out of one of your dojo videos (can't recall which one - and they are not searchable for code so I'd have to watch them till I find the one I want),,,,, Maybe that approach was valid back in the version 4.0.0 - RC days, but has been superseded now?

**Reply by ColinWPA** [reply]
Ah, passing the configuration and release in as command line parameters works (after taking the constant settings out of the project file. 
Great - progress at last! Thanks

**Reply by robmen** [reply, marked as answer]
At the end of each Deployment Dojo episode (once GitHub is introduced) the code is pushed to the [Deployment Dojo BeltTest repository](https://github.com/DeploymentDojo/BeltTest). You can see each episode in the history of the repo.
